---
title: "BIRSBIO2020.scSpatial.sPLS-DA+MINT"
author: "Yuzhou Feng"
date: "24/11/2020"
output: 
  pdf_document:
    toc: true    # can remove
    toc_depth: 3 # can remove
    number_sections: true
  highlight: zenburn
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, eval = TRUE)
```

```{r,echo=FALSE}
library(matrixStats)
library(mixOmics)
library(preprocessCore)
library(dplyr)
```

\bigskip

# Introduction

## Background 

Tasic et al. (2016) used scRNA sequencing to explore the extent of cell types of mouse primal visual cortex. In the study, 1723 high-quality cells with thousands of genes in primal visual cortex in male adult mice were sequenced on a single cell resolution for cell classification. This study identified 49 cell types, including 23 GABAergic, 19 glutamatergic and 7 non-neural types in the 1723 cells.

The cell classification result from Tasic et al. (2016) was used by Zhu et al. (2018) to design a study to distinguish the difference between intrinsic and extrinsic effect on gene expression. Intrinsic effect means the regulatory gene network, while extrinsic means the cellular microenvironment. The study was conducted by combining scRNA sequencing data and smFISH data (single-molecule fluorescence in situ hybridization). The former one has high molecular resolution of transcriptomics (thousands of genes) without spatial information, while the latter keeps the spatial information but loses the high resolution (only a few hundred genes). 

Zhu et al. (2018) mapped the sRNA sequencing data to the seqFISH data to enhance the molecular resolution of the cells with spatial information using SVM (support vector machine). The model was trained to identify the major cell type difference by training on scRNA data of 8 groups, GABAergic and Glutamatergic are the major neuron types, and other non-neuronal types, including Astrocytes, Mndothelial cells, microcytes, and three types of Oligocytes. The selected features (genes) were the top 43 differentially expressed (DE) genes in the identified 113 genes. The classification result was FURTHER validated by different evidence like cell type specific staining and previously reported marker genes.

Now, given a dataset of gene expression levels of 1723 cells and 113 genes and the cell types (from Tasic et al. 2016), our task is to classify another set of cells provided by Zhu et al. (2018), 1597 cells with 113 genes.

**Outline:**

I will adopt a semi-supervised approach to classify the cells of seqFISH data.

1. Train an sPLS-DA model on scRNA dataset. 
2. Select genes by limiting the number of genes of "keepX" argument of each component during hyperparameter tuning.
3. Predict seqFISH data using the trained sPLS-DA model. 
4. The predictions with high probabilities by the sPLS-DA model will be combined to the original training data for further training. 
5. Train a MINT model on the combined training dataset and predict the rest cells of seqFISH data.
6. Identify the marker genes by examining the loading factors.

This semi-supervised method can borrow the information from the seqFISH data and make the model customised. The combination of the data will be used to train a MINT.sPLS-DA model. MINT (Multivariate INTegrative method) (Rohart et al. 2017) is robust for integrating data from different sources regardless of the batch effect. The top discriminative genes identified from the model will be validated using previous literature as evidence. The determination of the minimal number of genes will be done by restricting the values of keepX and the performance will be monitored by balanced error rate (BER).

## Data

First import the data, including the training data (`scRNA`), the training labels (`scRNA_label`), the test data (`seqFISH`) and the predicted test labels by Zhu's paper (`seqFISH_label`).

```{r}
##### import the data #####
scRNA1 = read.delim("data/tasic_training_b2.txt",header = FALSE, sep ='\t') 
scRNA <- scRNA1[,-1]
rownames(scRNA) <- scRNA1[,1]
scRNA = data.frame(t(scRNA))
rownames(scRNA) <- paste(rep("scRNA",dim(scRNA)[1]),1:dim(scRNA)[1],sep = "_")
  
seqFISH1 = read.delim("data/seqfish_cortex_b2_testing.txt",header = FALSE, sep ='\t') 
seqFISH <- seqFISH1[,-1]
rownames(seqFISH) <- seqFISH1[,1]
seqFISH = data.frame(t(seqFISH))
rownames(seqFISH) <- paste(rep("seqFISH",dim(seqFISH)[1]),1:dim(seqFISH)[1],sep = "_")

scRNA_label = read.table("data/tasic_labels.tsv", sep = "\t")[,1]
seqFISH_label = read.table("data/seqfish_labels.tsv", sep = "\t")[,3]
```

`scRNA` is from Taisc et el. 2016. It contains the gene expression level of 1723 cells with 113 genes. 

```{r}
dim(scRNA)
head(scRNA[,1:6])
n_genes <- 113
n_scRNA <- 1723
```

`seqFISH` is from Zhu et el. 2018. It contains the gene expression level of 1597 cells with the 113 genes shared with `scRNA`. 

```{r}
dim(seqFISH)
head(seqFISH[,1:6])
n_seqFISH <- 1597
```

`scRNA_label` has the class of the cells corresponding to `scRNA` data. The phenotyping was done by Tasic et al. and the labels are treated as ground truth in our analysis and are used for training models.  

```{r}
length(scRNA_label)
table(scRNA_label)
```

There are 8 classes in `scRNA_label`. The numbers of cells in each type are not balanced.  

`seqFISH_label` has the classes predicted for `seqFISH` data by Zhu et al. The predicitons have not been benchmarked, thus `seqFISH_label` will not be used in our analysis.

```{r}
length(seqFISH_label)
table(seqFISH_label)
```

# Data preprocessing

## Quantile normalisation

First, we check the gene expression distribution of each gene in `scRNA` and `seqFISH`.

```{r, echo = FALSE, eval = FALSE}
for (i in 1:dim(scRNA)[2]){
  hist(scRNA[,i], main = colnames(scRNA)[i])
}
```

![](hist_sc1.png)
![](hist_sc2.png)

```{r, echo = FALSE, eval = FALSE}
for (i in 1:dim(seqFISH)[2]){
  hist(seqFISH[,i], main = colnames(seqFISH)[i])
}
```

![](hist_fish1.png)
![](hist_fish2.png)

The histograms show that the features (genes) are of the same scale and the distributions are almost normal. It means that there's no need for further feature normalisation.

Now check the gene expression distribution of the 40 random samples in each dataset. 

```{r,echo = FALSE, fig.height = 3, fig.width=8}
par(mfrow = c(1,1))
random_scRNA = sample(1:n_scRNA, 40, replace=FALSE)
boxplot(t(scRNA)[,random_scRNA], names = random_scRNA, 
        main = "gene expression boxplot of 40 random scRNA samples", xlab = "sample ID (scRNA)",cex.main = 1)
random_seqFISH = sample(1:n_seqFISH, 40, replace=FALSE)
boxplot(t(seqFISH)[,random_seqFISH], names = random_seqFISH, 
        main = "gene expression boxplot of 40 random seqFISH samples", xlab = "sample ID (seqFISH)",cex.main = 1)
par(mfrow = c(1,1))
```

Figure 1. Gene expression distribution of each sample in scRNA data and seqFISH data.

The boxplots show that the the distributions of scRNA data have some sample-wise variation, while seqFISH samples have more consistent distributions.

Some other analyses in the hackalthon did quantile normalisation for paired genes to try to eliminate the differences across datasets. However, I insist that we should do **quantile normalisation** on **samples**. The rationale behind sample quantile normalisation is that although different samples have different differentially expressed (DE) genes, the majority of the genes should express at the same level. That means whichever genes are DE, the mean and quantile of each sample should be of the similar level. 

Therefore, I will unify the quantile normalisation of samples between the datasets to maximumly eliminate the differences between the datasets. The normalisation was done by `normalize.quantiles` function from `preprocessCore` package.

```{r,fig.height = 6}
## QUANTILE normalisation on all the samples
all_data = rbind(scRNA,seqFISH)
qt_all_data = t(normalize.quantiles(t(all_data)))
colnames(qt_all_data) <- colnames(scRNA)
# get the quantile normalised scRNA and seqFISH data
qt_scRNA = data.frame(qt_all_data[1:dim(scRNA)[1],])
qt_seqFISH = data.frame(qt_all_data[(1+dim(scRNA)[1]):(dim(scRNA)[1]+dim(seqFISH)[1]),])
```

Then check 40 random samples from each dataset again.

```{r,echo = FALSE,fig.height = 3, fig.width=8}
par(mfrow = c(1,1))
random_qt_scRNA = sample(1:n_scRNA, 40, replace=FALSE)
boxplot(t(qt_all_data)[,random_qt_scRNA], names = random_qt_scRNA, 
        main = "gene expression boxplot of 40 random scRNA samples after qt normalisation",
        xlab = "sample ID (scRNA)",cex.main = 1)
random_qt_seqFISH = sample(1:n_seqFISH, 40, replace=FALSE)
boxplot(t(qt_all_data)[,random_qt_seqFISH], names = random_qt_seqFISH, 
        main = "gene expression boxplot of 40 random seqFISH samples after qt normalisation", 
        xlab = "sample ID (seqFISH)",cex.main = 1)
par(mfrow = c(1,1))
```

Figure 2. Gene expression distribution of 40 random samples in scRNA data and seqFISH data after quantile normalisation.

`qt_scRNA` is quantile normalised scRNA data and `qt_seqFISH` is quantile normalised seqFISH data.

## Oversampling

Let's check the number of samples in each class.

```{r}
table(scRNA_label)
```

The cells of each type are heavily imbalanced in the training data. There is not enough training data for the minor group and the major groups will dominate the classification. Then the model will not have enough power to classify minor groups. To solve the problem, I used two strategies. First, I combined the three subgroups of Oligodendroctyes into one group. This is because in the Tasic paper, some cells were not identified as "core" cells for any Oligo but identified as "intermediate" cells and these three types were identified as close to each other. Also, Oligo.2 has too few samples. The second strategy is to oversample the minor groups. 

First combine the three Oligo subgroups.

```{r}
scRNA_relabel = scRNA_label
scRNA_relabel[c(which(scRNA_relabel == "Oligodendrocyte.1"), which(scRNA_relabel == "Oligodendrocyte.2"), which(scRNA_relabel == "Oligodendrocyte.3"))] = "Oligodendrocyte"
table(scRNA_relabel)
```

```{r}
qt_scRNA_Astro = qt_scRNA[which(scRNA_label == "Astrocyte"),]
qt_scRNA_Endo = qt_scRNA[which(scRNA_label == "Endothelial Cell"),]
qt_scRNA_GABA = qt_scRNA[which(scRNA_label == "GABA-ergic Neuron"),]
qt_scRNA_Glu = qt_scRNA[which(scRNA_label == "Glutamatergic Neuron"),]
qt_scRNA_Micro = qt_scRNA[which(scRNA_label == "Microglia"),]
qt_scRNA_O1 = qt_scRNA[which(scRNA_label == "Oligodendrocyte.1"),]
qt_scRNA_O2 = qt_scRNA[which(scRNA_label == "Oligodendrocyte.2"),]
qt_scRNA_O3 = qt_scRNA[which(scRNA_label == "Oligodendrocyte.3"),]
qt_scRNA_Oligo = rbind(qt_scRNA_O1, qt_scRNA_O2,qt_scRNA_O3)
```

Now, oversample the minor groups.

```{r}
com_qt_scRNA_label = c(rep("Astrocyte", 43*15), rep("Endothelial Cell",29*25),rep("GABA-ergic Neuron",761), rep("Glutamatergic Neuron", 812), rep("Microglia",22*30),rep("Oligodendrocyte", 56*12))
com_qt_scRNA_label = as.factor(com_qt_scRNA_label)
```

We end up with around 700 samples in each class. The total number of cells after oversampling is 4275.

```{r}
length(com_qt_scRNA_label)
table(com_qt_scRNA_label)
```

Correspondingly, the `scRNA` data should also be resized. `com_qt_scRNA` is the oversampled scRNA data. It has 4275 samples.`com_qt_scRNA_label` is the new scRNA label vector.

```{r}
com_qt_scRNA_Astro = qt_scRNA_Astro %>% slice(rep(1:n(), each = 15))
com_qt_scRNA_Endo = qt_scRNA_Endo %>% slice(rep(1:n(), each = 25))
com_qt_scRNA_GABA = qt_scRNA_GABA 
com_qt_scRNA_Glu = qt_scRNA_Glu
com_qt_scRNA_Micro = qt_scRNA_Micro %>% slice(rep(1:n(), each = 30))
com_qt_scRNA_Oligo = qt_scRNA_Oligo %>% slice(rep(1:n(), each = 12))
com_qt_scRNA = rbind(com_qt_scRNA_Astro, com_qt_scRNA_Endo, com_qt_scRNA_GABA, com_qt_scRNA_Glu,
                     com_qt_scRNA_Micro, com_qt_scRNA_Oligo)
```


```{r}
dim(com_qt_scRNA)
```




\pagebreak
```{r}
sessionInfo()
```


